test/lib/jdk/test/lib/hexdump/HexPrinter.java

```diff
@@ -1185,5 +1185,12 @@ public int read(byte[] b, int off, int len) throws IOException {
             byteOffset += Math.max(size, 0);
             return size;
         }
+
+        @Override
+        public long transferTo(OutputStream out) throws IOException {
+            long size = super.transferTo(out);
+            byteOffset += size;
+            return size;
+        }
```

Comment:
This is an indication that overriding `transferTo()` in `BufferedInputStream` has potential compatibility impacts on its subclasses. Therefore I would suggest adding a check in `BufferedInputStream::transferTo` to only override the behavior if not subclassed. Something like:

```
     if (this.getClass() != BufferedInputStream.class) {
         return super.transferTo(out);
     }
     ... otherwise proceed with the new implementation ...
```

Then you can revert this change here.

Reply:
I do not see that it makes much sense to add such a safety means before I am finished with the inspection of the subclasses. As you can see in this example, it was pretty easy to fix it, and there are only few such subclasses at all. So instead of preparing against possibly non-existing easyt-to-fix problems, I prefer fixing the actual problems. They can only occur inside of the JDK itself, as relying on any particular implementation inside of the JDK by non-JDK classes simply would be a complete programming fault and *must* get fixed.

Reply:
It's a well known behavior that overriding the`read(...)` method is sufficient to implement subclass behavior. This will no longer be the case if `transferTo(...)` no longer calls `this.read(...)` as it used to. There are many undocumented behaviors that have been depended on over the year - and though I agree with the sentiment that it's not a good idea to depend on undocumented behavior, this is one that is long standing, and mostly legitimate, especially for those classes that were coded before `transferTo` was added to `InputStream`.

The issue here is that things would mostly work for those custom subclassed, up to the point where some unsuspecting code might call `transferTo`. This could lead to bugs that might become hard to diagnose, and of course would only be seen at runtime.
As a point of comparison, look at what [JEP 425](https://openjdk.org/jeps/425) did to change the locking in `BufferedInputStream` (see how the lock object is used/initialized).

How many classes of `BufferedInputStream` are there in the wild, outside of the JDK?
How many of them have custom behavior implemented in `read`, that would be broken if `transferTo` no longer called `this.read`?

If you're changing an observable behavior, you will need a CSR, you will need to evaluate the compatibility risk and justify it, you will need to write a Release Note to warn about the compatibility risk, so that custom implementations that have such a dependency can be changed to also override `transferTo`.

The main question is whether the advantage of changing the behavior outweigh the risk of regression it might cause in subclasses. Such might be the case - or not.
The middle ground - and safer path - is to override the behaviour only for the base class and those subclasses that do not depend on the default behavior as was done by JEP 425...

Reply:
> It's a well known behavior that overriding the`read(...)` method is sufficient to implement subclass behavior. This will no longer be the case if `transferTo(...)` no longer calls `this.read(...)` as it used to. There are many undocumented behaviors that have been depended on over the year - and though I agree with the sentiment that it's not a good idea to depend on undocumented behavior, this is one that is long standing, and mostly legitimate, especially for those classes that were coded before `transferTo` was added to `InputStream`.
> 
> The issue here is that things would mostly work for those custom subclassed, up to the point where some unsuspecting code might call `transferTo`. This could lead to bugs that might become hard to diagnose, and of course would only be seen at runtime. As a point of comparison, look at what [JEP 425](https://openjdk.org/jeps/425) did to change the locking in `BufferedInputStream` (see how the lock object is used/initialized).
> 
> How many classes of `BufferedInputStream` are there in the wild, outside of the JDK? How many of them have custom behavior implemented in `read`, that would be broken if `transferTo` no longer called `this.read`?
> 
> If you're changing an observable behavior, you will need a CSR, you will need to evaluate the compatibility risk and justify it, you will need to write a Release Note to warn about the compatibility risk, so that custom implementations that have such a dependency can be changed to also override `transferTo`.
> 
> The main question is whether the advantage of changing the behavior outweigh the risk of regression it might cause in subclasses. Such might be the case - or not. The middle ground - and safer path - is to override the behaviour only for the base class and those subclasses that do not depend on the default behavior as was done by JEP 425...

I understand your claim but disagree with the conclusion. You could impose it to *any* behavior change of *any* class -- this is no special case here. I do not see that we need a CSR as I do not change the *documented* behavior (neither the JavaDoc nor the Spec). Also I need to ask back: Do you *really* assume there are such lots of people using `transferTo` with failing subclasses of BIS? I doubt that, actually. You try to fix things possible not broken at all. We should concentrate on facts, not on assumptions, IMHO.

Having said that, I leave it up to @AlanBateman to decide, as I have no strong feelings about the issue you raised.

Reply:
With the changes you proposed a CSR will definitely be needed.

Reply:
BIS dates from JDK 1.0 and it's not hard to find examples that extend it. The HexPrinter test reminds us that it has been possible for 25+ years to override the read methods and snoop on all bytes that are read. Adding an override of the transferTo method that bypasses the read methods might break these subclasses. So I think Daniel is right that we have to be cautious.

So I think the feedback that you have now is that bypassing the buffer is problematic when using mark, when there are buffered bytes, or when BIS has been extended. It may be that the starting point is something very conservative like this:

```
    @Override
    public long transferTo(OutputStream out) throws IOException {
        if (getClass() == BufferedInputStream.class
                && ((count - pos) <= 0) && (markpos == -1)) {
            return getInIfOpen().transferTo(out);
        } else {
            return super.transferTo(out);
        }
    }
```

There is also the issue of locking and async close to look into.

Reply:
Thank you for your decision about this issue, so instead of fixing the existing bugs I will implement as you outlined.

BTW, locking and rebasing is on the way.

Reply:
Rebased and fixed locking, using your proposed code now.

Async close leaves BIS in an invalid state. The JavaDocs say ` The behavior for the case where the inputand/or output stream is asynchronously closed, or the thread interrupted during the transfer, is highly input and output streamspecific, and therefore not specified.`. So there is seems to be no *essential* need to do something *particular* in that case (the caller cannot further use the BIS, just like after an IOE: `It is strongly recommended that both streams be promptly closed if an I/O error occurs.` -- which is what *I personally, as a user of BIS* would do after async close), but if you *want* that we reach a particular outcome, then just tell me what outcome you envision.

Reply:
> With the changes you proposed a CSR will definitely be needed.

Just to learn about OpenJDK rules: Why? I always thought that CSR is only needed when breaking *documented* behavior. Do you have a link to the official rule that proofs your claim in *this* case (i. e. *non-documented* behavior)? Thanks! :-)

Reply:
Hi Markus, It's a good idea to write a CSR if your change modifies an observable behavior that users of the public API (including custom subclasses) might be depending on. This is both to raise the awareness on the impact of the change, document the possible regression risks, and get feedback from the CSR committee who might suggest mitigating actions to reduce the risk (such as e.g: adding a property to make it possible to revert to the old behavior, write a release note, etc...). In the case at hand `BufferedInputStream` is a very old class that has been used and abused all over the board in the java ecosystem. So writing a CSR to document a change in behavior that some subclasses may be depending on would be the proper thing to do, when changing that behaviour. The fact that you have found such a subclass (with a legitimate use case) only strengthen the need for exercising such care.
In retrospect it would have been much better if `BufferedInputStream` had been declared final - but it's much too late for that now. Thanks!

Reply:
> @AlanBateman Async close leaves BIS in an invalid state. The JavaDocs say ` The behavior for the case where the input and/or output stream is asynchronously closed, or the thread interrupted during the transfer, is highly input and output stream specific, and therefore not specified.`. 

Yes, I believe I suggested that wording when the method was added. The java.io APIs mostly pre-date deep consideration of those topics.

I've looked at the update change to BIS in f2f9a904 and I think it's okay.

On the CSR topic and your messages with Daniel. On the CSR wiki pages then phrase to look for is "behavioral compatibility".  The "Kinds of Compatibility" has more details on the topic.

Reply:
Thank you all for your kind help and feedback.

@AlanBateman So is now the time to switch this PR from Draft to Ready?

Reply:
The source changes LGTM. I suppose you should now revert the changes above (in `HexPrinter.java`), otherwise the bytes will now be counted twice :-).
Which BTW shows that you would be depending on undocumented behavior whatever you do:
`transferTo` is not documented to call `this.read(...)`, but neither is it documented to **not call**  `this.read(...)`...  Hyrum's law wins here ;-)

Reply:
https://github.com/openjdk/jdk/pull/6935/commits/b4d5004dd3299d084769309b505b19ed7dd3182e reverts the change, as it is not needed (and wrong) with the check in place. I wonder if in future people will actually understand what actually is going on under the hood here.

Reply:
> @AlanBateman So is now the time to switch this PR from Draft to Ready?

Yes. From what I can tell, moving to Draft while you worked on it didn't pause the Skara bots, I'm not sure why that is.